specs/kyber/src/sampling.rs

```diff
@@ -5,67 +5,78 @@ use crate::{
-///     if d_1 < q then
-///         aˆ_{j} := d_1
-///         j := j + 1
+/// Input: byte stream B ∈ B*
+/// Output: array â ∈ ℤ₂₅₆
+///
+/// i ← 0
+/// j ← 0
+/// while j < 256 do
+///     d₁ ← B[i] + 256·(B[i+1] mod 16) d2 ← ⌊B[i+1]/16⌋+16·B[i+2] ifd1 <qthen
+///     d₂ ← ⌊B[i+1]/16⌋ + 16·B[i+2]
+///     if d₁ < q then
+///         â[j] ← d₁
+///         j ← j + 1
 ///     end if
-///     if d_2 < q and j < n then
-///         aˆ_{j} := d_2
-///         j := j + 1
+///     if d₂ < q and j < 256 then
+///         â[j] ← d₂
+///         j ← j+1
 ///     end if
-///     i := i + 3
+///     i ← i+3
 /// end while
-/// return aˆ0 + aˆ1X + · · · + aˆ{n−1}X^{n−1}
-/// ```
+/// return â
 ///
-/// The Kyber Round 3 specification can be found at:
-/// <https://pq-crystals.org/kyber/data/kyber-specification-round3-20210131.pdf>
-pub fn sample_from_uniform_distribution(
-    randomness: [u8; parameters::REJECTION_SAMPLING_SEED_SIZE],
+/// The NIST FIPS 203 standard can be found at
+/// <https://csrc.nist.gov/pubs/fips/203/ipd>.
+pub fn sample_ntt(
+    bytes: [u8; parameters::REJECTION_SAMPLING_SEED_SIZE],
 ) -> Result<KyberPolynomialRingElement, BadRejectionSamplingRandomnessError> {
     let mut sampled_coefficients: usize = 0;
-    let mut out: KyberPolynomialRingElement = KyberPolynomialRingElement::ZERO;
+    let mut a_hat: KyberPolynomialRingElement = KyberPolynomialRingElement::ZERO;
 
-    for bytes in randomness.chunks(3) {
-        let b = u16::from(bytes[0]);
-        let b1 = u16::from(bytes[1]);
-        let b2 = u16::from(bytes[2]);
+    for byte_chunk in bytes.chunks(3) {
+        let b = u16::from(byte_chunk[0]);
+        let b1 = u16::from(byte_chunk[1]);
+        let b2 = u16::from(byte_chunk[2]);
 
+        // d_1 ← B[i] + 256·(B[i+1] mod 16)
+        // d_2 ← ⌊B[i+1]/16⌋ + 16·B[i+2]
         let d1 = b + (256 * (b1 % 16));
-
-        // Integer division is flooring in Rust.
         let d2 = (b1 / 16) + (16 * b2);
 
-        if d1 < parameters::FIELD_MODULUS && sampled_coefficients < out.len() {
-            out[sampled_coefficients] = d1.into();
+        // if d_1 < q then
+        //     â[j] ← d_1
+        //     j ← j+1
+        // end if
+        if d1 < parameters::FIELD_MODULUS && sampled_coefficients < a_hat.len() {
+            a_hat[sampled_coefficients] = d1.into();
             sampled_coefficients += 1
         }
-        if d2 < parameters::FIELD_MODULUS && sampled_coefficients < out.len() {
-            out[sampled_coefficients] = d2.into();
+
+        // if d_2 < q and j < 256 then
+        //     â[j] ← d_2
+        //     j ← j+1
+        // end if
+        if d2 < parameters::FIELD_MODULUS && sampled_coefficients < a_hat.len() {
+            a_hat[sampled_coefficients] = d2.into();
             sampled_coefficients += 1;
         }
 
-        if sampled_coefficients == out.len() {
-            return Ok(out);
+        if sampled_coefficients == a_hat.len() {
+            return Ok(a_hat);
```

Comment:
This will make trouble in hax because of early returns. I have to figure out if we can get this fixed in the tool or should work around for now.

Reply:
Moved the return out of the loop.